pulpcore/pulpcore/tasking/worker.py

```diff
@@ -0,0 +1,119 @@
+import logging
+import os
+import signal
+import socket
+import sys
+import threading
+import time
+
+from rq import Queue
+from rq.worker import Worker
+
+
+import django  # noqa otherwise E402: module level not at top of file
+django.setup()  # noqa otherwise E402: module level not at top of file
+
+
+from pulpcore.app.models import Task
+
+from .constants import TASKING_CONSTANTS
+from .services.storage import WorkerDirectory
+from .services.worker_watcher import (check_worker_processes, handle_worker_heartbeat,
+                                      mark_worker_offline)
+
+
+_logger = logging.getLogger(__name__)
+
+
+class PulpWorker(Worker):
+    """
+    A Pulp worker for both the resource manager and generic workers
+
+    This worker is customized in the following ways:
+
+        * Replaces the string '%h' in the worker name with the fqdn
+        * If the name starts with 'reserved_resource_worker' the worker ignores any other Queue
+          configuration and only subscribes to a queue of the same name as the worker name
+        * If the name starts with 'resource_manager' the worker ignores any other Queue
+          configuration and only subscribes to the 'resource_manager' queue
+        * Sets the worker TTL
+        * Supports the killing of a job that is already running
+    """
+
+    # Do not print "Result is kept for XXX seconds" after each job
+    log_result_lifespan = False
+
+    def __init__(self, queues, **kwargs):
+
+        kwargs['name'] = kwargs['name'].replace('%h', socket.getfqdn())
+
+        if kwargs['name'].startswith(TASKING_CONSTANTS.WORKER_PREFIX):
+            queues = [Queue(kwargs['name'], connection=kwargs['connection'])]
+        if kwargs['name'].startswith(TASKING_CONSTANTS.RESOURCE_MANAGER_WORKER_NAME):
+            queues = [Queue('resource_manager', connection=kwargs['connection'])]
+
+        kwargs['default_worker_ttl'] = TASKING_CONSTANTS.WORKER_TTL
+
+        return super().__init__(queues, **kwargs)
+
+    def perform_job(self, job, queue):
+        """
+        Set the :class:`pulpcore.app.models.Task` object in the running state.
+
+        This method is called by the worker just before the task begins executing.
+        """
+        try:
+            task = Task.objects.get(pk=job.get_id())
+        except Task.DoesNotExist:
+            pass
+        else:
+            task.set_running()
+
+        def check_kill(conn, id, interval=1):
+            while True:
+                res = conn.srem(TASKING_CONSTANTS.KILL_KEY, id)
+                if res > 0:
+                    os.kill(os.getpid(), signal.SIGKILL)
+                time.sleep(interval)
+
+        t = threading.Thread(target=check_kill, args=(self.connection, job.get_id()))
+        t.start()
+
+        return super().perform_job(job, queue)
+
+    def handle_job_failure(self, job, **kwargs):
```

Comment:
These next couple of methods need docstrings

Reply:
fixed in next push, ty.